File: scripts/oop-practice-sources/php/animal-inheritance.php
<?php

class Animal
{
    public function __construct(protected string $name) {}

    public function eat(): void
    {
        echo "{$this->name} ест\n";
    }
}

class Cat extends Animal
{
    public function speak(): void
    {
        echo "Мяу!\n";
    }
}

class Dog extends Animal
{
    public function speak(): void
    {
        echo "Гав!\n";
    }
}

$cat = new Cat('Мурка');
$dog = new Dog('Шарик');
$cat->eat();
$cat->speak();
$dog->eat();
$dog->speak();

File: scripts/oop-practice-sources/lua/animal-inheritance.lua
local Animal = {}

function Animal:new(name)
    local obj = setmetatable({name = name}, {__index = self})
    return obj
end

function Animal:eat()
    print(self.name .. " ест")
end

local Cat = {}
setmetatable(Cat, {__index = Animal})

function Cat:new(name)
    return Animal.new(self, name)
end

function Cat:speak()
    print("Мяу!")
end

local Dog = {}
setmetatable(Dog, {__index = Animal})

function Dog:new(name)
    return Animal.new(self, name)
end

function Dog:speak()
    print("Гав!")
end

local cat = Cat:new("Мурка")
local dog = Dog:new("Шарик")
cat:eat()
cat:speak()
dog:eat()
dog:speak()
